pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -246,13 +255,18 @@ public void removeSegment(IndexSegment segment) {
     LOGGER.info("Removing upsert metadata for segment: {}", segmentName);
 
     if (!Objects.requireNonNull(segment.getValidDocIds()).getMutableRoaringBitmap().isEmpty()) {
-      // Remove all the record locations that point to the removed segment
-      _primaryKeyToRecordLocationMap.forEach((primaryKey, recordLocation) -> {
-        if (recordLocation.getSegment() == segment) {
-          // Check and remove to prevent removing the key that is just updated
-          _primaryKeyToRecordLocationMap.remove(primaryKey, recordLocation);
-        }
-      });
+      PrimaryKey reuse = new PrimaryKey(new Object[]{});
```

Comment:
We want to pre-allocate the buffer so that it can be reused
```suggestion
      PrimaryKey reuse = new PrimaryKey(new Object[_primaryKeyColumns.size()]);
```